### DISCUSSION TITLE
Zoom Tool behaviour

### DISCUSSION THREAD

**Discussion by novhack** [Orama-Interactive/Pixelorama#540] [Ideas]
Currently the Zoom Tool has a mode selector - either zoom in or zoom out. I propose a change to how it behaves because I don't think the current workflow is very usable.

The way how it works in Aseprite seems like a better approach:
`When you click and drag to left you zoom out from the pixel you clicked on. When you drag to right you zoom in.`

This seems much more suited for graphics tablet users than clicking on the selector every time (I expect that most mouse and touchpad users are using scrolling anyway).

**Comment by Variable-ind** [comment]
That feature could be useful :+1: . Well, In the mean time you can also set custom zoom level by double clicking/tapping the zoom label

**Comment by mrtripie** [comment]
I was experimenting with a different approach a few weeks ago, based on how 3D software tend to approach it:

https://user-images.githubusercontent.com/65431647/134386310-4b86665a-3377-4986-9597-78830327b72f.mp4

The idea is to be able to avoid switching tools entirely.

The branch is here if you want to see how it feels: https://github.com/mrtripie/Pixelorama/tree/NavigationIcons

**Reply by Variable-ind** [reply]
Oh!!, i was just wondering about that! so it's finally done!?

**Reply by mrtripie** [reply]
No, I put it off to work on my own project when the focus was on finalizing 2.9. Its usable, but the rotation aspect needs to be ironed out, along with maybe some other tweaks + clean up.

**Reply by Variable-ind** [reply]
Just tried it out, It looks great :+1:!, However i would also suggest some sort of reset rotation option since i wasn't able to adjust it's rotation back to zero degrees (it always got ahead or back a few degrees).

**Reply by mrtripie** [reply]
Right clicking on each of the options should reset them.

I was wondering if it might be better to use something like double click though as someone working on a touch screen may not be able to do right click? (don't have any touch screen PCs so not sure)

**Reply by Variable-ind** [reply]
it may work with only one click (a timer with a given limit say 0.3 seconds starts as soon as we press the mouse button. if the click is released while the timer is still active then it resets the values else it goes into drag mode)

**Reply by mrtripie** [reply]
I would think using a mouse movement threshold would be better (save a variable of the start mouse position, and if its moved say 2 pixels, go into a navigation state)

**Reply by mrtripie** [reply]
One thing the rotation needs is the degrees displayed at the top, similar to zoom.

I was thinking of changing both rotation and zoom to a button so that it gets a highlight when hovered, I also added the finger mouse cursor and making it just single click rather than double, as I think the feature wasn't very discoverable before:

https://user-images.githubusercontent.com/65431647/134592147-1f803e52-f965-4372-91d7-fc3f9d23cee7.mp4

The layer names also don't show an indication that they can be clicked on, so maybe it makes sense to do something similar there too (although its something you would assume you could do if you've used similar apps)

I also thought it may be interesting if clicking the canvas size label would bring up resize canvas.

**Reply by Variable-ind** [reply]
Looks great, but i don't think doing this with canvas is a good idea cause there are two dialogues for resizing... one for resizing only canvas and other for scaling image along with canvas resize

**Comment by OverloadedOrama** [comment]
I went ahead and implemented the feature in commit 77996c769e2a1bcbdc911882d14359da4eaec8b3. I also kept the original behavior that zooms in and out on a single click, but I made it so that if <kbd>Control</kbd> is held, the mode is changed, as suggested in #429. Let me know if it works well!
